src/audio/dp_queue.c

```diff
@@ -0,0 +1,328 @@
+
+	struct k_spinlock lock;
+};
+
+static inline bool dp_queue_is_shared(struct dp_queue *dp_queue)
+{
+	return !!(dp_queue->flags & DP_QUEUE_MODE_SHARED);
+}
+
+static inline uint8_t *dp_queue_buffer_end(struct dp_queue *dp_queue)
+{
+	return dp_queue->data_buffer + dp_queue->data_buffer_size;
+}
+
+struct dp_queue *dp_queue_get_from_sink(struct sof_sink __sparse_cache *sink)
+{
+	struct dp_queue_sink_source_c __sparse_cache *sink_src_c =
+		attr_container_of(sink, struct dp_queue_sink_source_c __sparse_cache,
+				  sink, __sparse_cache);
+
+	return sink_src_c->owner;
+}
+
+struct dp_queue *dp_queue_get_from_source(struct sof_source __sparse_cache *source)
+{
+	struct dp_queue_sink_source_c __sparse_cache *sink_src_c =
+		attr_container_of(source, struct dp_queue_sink_source_c __sparse_cache,
+				  source, __sparse_cache);
+	return sink_src_c->owner;
+}
+
+static inline void dp_queue_invalidate_shared(struct dp_queue *dp_queue,
+					      void __sparse_cache *ptr, uint32_t size)
+{
+	/* rollback? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_invalidate_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+	}
+	/* invalidate rest of data - from begin of buffer, remaining size */
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)dp_queue->data_buffer, size);
+}
+
+static inline k_spinlock_key_t dp_queue_lock(struct dp_queue *dp_queue)
+{
+	return k_spin_lock(&dp_queue->lock);
+}
+
+static inline void dp_queue_unlock(struct dp_queue *dp_queue, k_spinlock_key_t key)
+{
+	k_spin_unlock(&dp_queue->lock, key);
+}
+
+struct sof_sink __sparse_cache *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->sink;
+}
+
+struct sof_source __sparse_cache *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->source;
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink __sparse_cache *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue->free_space;
+}
+
+static int dp_queue_get_buffer(struct sof_sink __sparse_cache *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	/* dp_queue_get_free_size will return free size with adjustment for cacheline if needed */
+	if (req_size > dp_queue->ibs || req_size > dp_queue_get_free_size(sink))
+		return -ENODATA;
+
+	/* no need to lock, just reading data that may be modified by commit_buffer only */
+	*data_ptr = dp_queue->data_buffer + dp_queue->write_offset;
+	*buffer_start = dp_queue->data_buffer;
+	*buffer_size = dp_queue->data_buffer_size;
+
+	/* provided buffer is an empty space, the requestor will perform write operations only
+	 * no need to invalidate cache - will be overwritten anyway
+	 */
```

Comment:
so is the buffer accessed via cached addresses? Then you need `__sparse_cache` for all those addresses. I think both for locking and cache synchronisation the reader will have to do the hard part. But this is the same situation here now as with the current buffer: currently your DP write and read pointers are uncached. So, frequently reading and writing them by both the reader and the writer will be slow. And if you make it cached you'll have to synchronise